### DISCUSSION TITLE
Problems with redirection in frontend after close/cancel when SEF is enabled

### DISCUSSION THREAD

**Discussion by fred-the-coder** [joomengine/Joomla-Component-Builder#908] [Q&A]
Hi all,
I saw many discussions around redirection, but did not find a solution to my problem...
My problem is most propably not related to JCB but I try my question here...

I have a JCB component that manages records.
On the web site I have a menu item that targets the list view of the records (`view=records`), let say menu item is Itemid=1.
I have another menu item that targets the edit view of a record (to be able to create one record): `view=record&layout=edit`, let say menu item is Itemid=2.
In this edit view I have the standard buttons SAVE/SAVE&CLOSE/SAVE&NEW/CANCEL.

In the edit view, to create a record, I have two differents behaviours when I clic on SAVE&CLOSE (or CANCEL) button:

- when SEF is disabled, the URL of my edit page is view=record&layout=edit&Itemid=2 and I am redirected to the list view (view=records), even if the URL ends like this `view=records&Itemid=2`. Where Itemid=2 is the menu item to the edit view...
- when SEF is enabled, the URL of my edit page is `/add-a-record` and I am redirected to the "same" edit view `/add-a-record`

I would like to be redirected to the list view when clicking on SAVE&CLOSE or CANCEL, as when SEF is disabled...

I have used the debugger to figure out why... but I find that a "return URL" may has to be set...and there I may try #816 or #870. But why? as it is working as I expect when SEF is disabled...

Any clue on this?
Thank you for your help!

N.B.: I saw that com_content Joomla component is now providing a Submission/Cancel Redirect value as option in menu items...

**Comment by cpaschen** [comment]
Did you take a look at Llewellyn's explanation of the routing that happens related to the buttons? :
https://github.com/vdm-io/Joomla-Component-Builder/discussions/873

I know it doesn't specifically explain your issue, but it did help give me an extremely better understanding how the routing works (or doesn't). And my guess is that the SEF issue is in some way related.

**Reply by fred-the-coder** [reply]
hi @cpaschen thanks for your reply.
OK I may add some custom code to force redirection as I want....

I will wait for another answer that could explain the difference of behaviour with/without SEF....to proceed.
I am afraid it is a Joomla rooting behaviour...

**Comment by Llewellynvdm** [comment]
There are changes in the core of Joomla regarding SEF that you will see in the article global options you can make a selection between the two under the integration tab:
![image](https://user-images.githubusercontent.com/5607939/165289548-e473b053-cb7c-4bfc-9798-cb7ce100fea9.png)

JCB just has the old legacy approach in place, and is not ideal. We have been looking to fix this, and move over to the newer better implementation, but so far we have just made a partial change. This means we are already doing things better, but there is still much room... currently we have made the choice to aim all our attention to getting JCB over to Joomla 4 and we are not actively looking to making any major improvements to JCB in Joomla 3.

This does not directly answer your why... but as to the technical reason you should study the router, as its behavior with SEF turned on... does change, and can be part of the reasons to why the outcome is not the same.

Redirection in Joomla is not a simple topic as there are so many factors and yet we have tried to cover as much edge cases as we could at the time. Going to Joomla 4 we will have the joy of all things new :+1:

**Reply by fred-the-coder** [reply]
Thanks @Llewellynvdm 
Depending on time available, I will follow in debug mode the step by step Joomla route construction...
I already reached the Joomla createURL function...maybe I am not that far.
Else I will go for the custom code approach...
